bench: replace wall-clock timer with per-process CPU timer

Raimo33 opened the pull request:
## Goal
This PR refactors the benchmarking functions as per #1701, in order to make benchmarks more deterministic and less influenced by the environvment. 

This is achieved by replacing Wall-Clock Timer with Per-Process CPU Timer when possible.

Raimo33 commented:
I think `min` and `max` just complicate things. let me explain:
first of all, as it is right now, they don't even measure the min and max, they just measure the min/max of the averages of all runs. aka not the absolute. Furthermore, in order to have them, one would need to run all the iterations 10 times more. benchmarks are already slow, adding this min/max slows them by 10 fold. imho it's completely unnecessary. @real-or-random

Raimo33 commented:
> I think there's a reason to have this. Some benchmarks take much longer than others, so it probably makes sense to run fewer iters for these.

right now all benchmarks are run with count=10 and fixed iters (apart from ecmult_multi which adjusts the number of iters, not count)

Raimo33 commented:
> Concept NACK on removing any ability to observe variance in timing. The current min/avg/max are far from perfect, but they work fairly well in practice. Improving is welcome, but removing them is a step backwards.

what is the usefulness of measuring min/max when we are removing OS interference & thermal throttling out of the equation? min/max will be extremely close to the avg no matter how bad the benchmarked function is.

Raimo33 commented:
by the way, `gettimeofday()` is officially discouraged since 2008 in favor of `clock_gettime()`. The POSIX standard marks it as **obsolescent** but still provides the API for backward compatibility.

Raimo33 commented on src/bench.h:
```diff
@@ -7,30 +7,63 @@
+#if defined(_WIN32)
+# include <windows.h>
+#else
+# include <time.h>
+# include <sys/time.h>
```
yes, one for `gettimeofday()` and one for `clock_gettime()`. they don't conflict.

Raimo33 replied:
Agree. Removed.

Raimo33 replied:
just tried. unfortunately the macros to detect wether `gettimeofday()` or `clock_gettime()` is needed are defined in the `time.h` header. so the only viable solution is:

```c
#include <time.h>
#if !defined(CLOCK_PROCESS_CPUTIME_ID) && !defined(CLOCK_MONOTONIC) && !defined(CLOCK_REALTIME)
#  include <sys/time.h>
#endif
```

which every LLM tells me is 'inherently unreliable' and can lead to portability issues.

> 'Including sys/time.h on modern POSIX systems that already have time.h and clock_gettime doesn’t hurt, and ensures gettimeofday is available if needed.'

> 'Trying to skip it adds complexity with almost no benefit.'

Raimo33 replied:
You're right, the latest specification makes `CLOCK_MONOTONIC` support mandatory, this makes it way simpler. no need to include `<sys/time.h>` anymore.

Raimo33 replied:
This would be possible if every compiler supported the Std 1003.1-2024 version of POSIX, which is unrealistic. Until then, for backward compatibility, I kept CLOCK_REALTIME as the last fallback.

Raimo33 replied:
and btw, the `_POSIX_CPUTIME` and `_POSIX_MONOTONIC` flags are to be used when you want to ensure that the library has mandatory support for those features. In my case for example, gcc doesnt expose `_POSIX_MONOTONIC` but still supports it. Therefore it's better to check `CLOCK_PROCESS_CPUTIME_ID` and `CLOCK_MONOTONIC` directly.

Raimo33 commented on src/bench.h:
```diff
@@ -7,30 +7,63 @@
+        return 0;
+    }
+    return (int64_t)ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
+# else
```
yes, I tried everything you mentioned.

removing redundancy here can be done by just using the ifdefs for the constant, but you'd need a different path for the last fallback. the last case uses tv, not ts.

I tried issuing warnings with `#warning` to tell the user about the precision of the clock but it's not compatible with c90.

Raimo33 replied:
Here's the DRY'd version:

```c
# if defined(CLOCK_PROCESS_CPUTIME_ID)
    /* in theory, CLOCK_PROCESS_CPUTIME_ID is only useful if the process is locked to a core. see https://linux.die.net/man/3/clock_gettime */
const clockid_t clk_id = CLOCK_PROCESS_CPUTIME_ID;
# elif defined(CLOCK_MONOTONIC)
    /* WARN: timer is influenced by environvment (OS, scheduling, interrupts...) */
const clockid_t clk_id = CLOCK_MONOTONIC;
# elif defined(CLOCK_REALTIME)
    /* WARN: timer is influenced by environvment (OS, scheduling, interrupts...) */
const clockid_t clk_id = CLOCK_REALTIME;
# else
# define USE_GETTIMEOFDAY
# endif

# ifndef USE_GETTIMEOFDAY
struct timespec ts;
if (clock_gettime(clk_id, &ts) != 0) {
    return 0;
}
return (int64_t)ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
# else
    /* WARN: timer is influenced by environvment (OS, scheduling, interrupts...) */
struct timeval tv;
gettimeofday(&tv, NULL);
return tv.tv_usec + (int64_t)tv.tv_sec * 1000000;
# endif
```

less readable imo

Raimo33 commented on src/bench.h:
```diff
@@ -7,92 +7,64 @@
-    /* C11 way to get wallclock time */
-    struct timespec tv;
-    if (!timespec_get(&tv, TIME_UTC)) {
-        fputs("timespec_get failed!", stderr);
-        exit(EXIT_FAILURE);
+    if (GetProcessTimes(GetCurrentProcess(), &creation, &exit, &kernel, &user) == 0) {
+        return 0;
```
previously wall-clock times.
now per-process cpu time.

Raimo33 commented on src/bench.h:
```diff
@@ -7,30 +7,63 @@
-    gettimeofday(&tv, NULL);
-    return (int64_t)tv.tv_usec + (int64_t)tv.tv_sec * 1000000LL;
+    gettimeofday((struct timeval*)&tv, NULL);
+    return (int64_t)tv.tv_sec * 1000000 + tv.tv_usec;
```
you're right. removed.

l0rinc commented on README.md:
```diff
@@ -159,6 +159,8 @@ Benchmark
 ------------
 If configured with `--enable-benchmark` (which is the default), binaries for benchmarking the libsecp256k1 functions will be present in the root directory after the build.
 
+For an accurate benchmark, it is strongly recommended to pin the process to a dedicated CPU core and to disable CPU frequency scaling.
```
This will be more reliable at best, but not more "accurate", since it doesn't reflect actual usage.

l0rinc commented:
I haven't investigated this in detail but couldn't find any related issue: were there any efforts to use https://nanobench.ankerl.com like we do in Core instead? That will likely need some adjustments on the benchmarks themselves, but at least we wouldn't be rolling our own framework.

Raimo33 commented:
I changed windows clock to `QueryPerformanceCounter()` which has microsecond precision.
I also removed if checks from `clock_gettime()` as it can only fail for programming errors.

l0rinc commented:
> but adding C++ to the code base seems a bit overkill to me if if's "just" for the purpose of benchmarks

The purpose would be to standardize the benchmarking to avoid reinventing solutions to problems that aren't strictly related to the purpose of this library. It would also help with trusting the result more, since we already know how that benchmarking library behaves.
The main client of this library is C++ code, it doesn't seem far-fetched to me to test that.

hebasto commented on src/bench.c:
```diff
@@ -4,6 +4,8 @@
  * file COPYING or https://www.opensource.org/licenses/mit-license.php.*
  ***********************************************************************/
 
+#define _POSIX_C_SOURCE 200112L /* for clock_gettime() */
```
Are there any references explaining why the value `200112L` was chosen?

Raimo33 replied:
no particular reason besides the fact that it is the bare minimum version to make it work on my machine and CI.
The reported `199309L` version number didn't expose the necessary macros.

hebasto replied:
> The reported `199309L` version number didn't expose the necessary macros.

I can't reproduce this on my machine.

> no particular reason besides the fact that it is the bare minimum version to make it work on my machine...

Could you share your build environment details so I can try to reproduce the issue?

> ... and CI.

It seems that CI is fine with the documented `199309L` value.
See: https://github.com/hebasto/secp256k1/actions/runs/18008380590

hebasto reviewed (commented):
According to https://www.man7.org/linux/man-pages/man3/clock_gettime.3.html:
> Link with `-lrt` (only for glibc versions before 2.17).

I think this check, and adding the `-lrt` flag if necessary, should be included in both build systems.

hebasto commented:
Friendly ping @martinus, benchmark connoisseur :)

Could you please give a rough estimate of these changes at the concept level?

Raimo33 commented on src/bench.h:
```diff
@@ -176,6 +190,14 @@ static int get_iters(int default_iters) {
     }
 }
 
+static void print_clock_info(void) {
+#if defined(CLOCK_PROCESS_CPUTIME_ID)
+    printf("INFO: Using per-process CPU timer\n\n");
+#else
+    printf("WARN: Using wall-clock timer instead of per-process CPU timer.\n\n");
+#endif
+}
+
```
while on windows there's no option for per-process cpu clock (or at least not high precision), I still think issuing the warning is fine.

Raimo33 commented:
Fixed by unconditionally adding `-lrt` when available.
```
root@5997b2bd1aca:/workspace# ldd --version | head -1
ldd (Ubuntu EGLIBC 2.15-0ubuntu10.23) 2.15
root@5997b2bd1aca:/workspace# cmake --build build -t bench
[ 33%] Built target secp256k1_precomputed
[ 66%] Built target secp256k1
[100%] Built target bench
```

purpleKarrot commented on src/CMakeLists.txt:
```diff
@@ -125,12 +125,17 @@ endif()
+  if(RT_LIBRARY)
+    target_link_libraries(optional_rt INTERFACE rt)
+  endif()
   add_executable(bench bench.c)
```
Note that CMake requires that

* test names are unique *per directory*
* target names are unique.

While it is not strictly required to prefix test names with the project, it is still recommended, because it allows users to control which tests to run: `ctest -R secp256k1_`. This recommendation is already followed, good.

For target names, this requirement is much stronger. Since this project is used as a subproject in bitcoin, neither bitcoin nor any other of its subprojects can define a target with the name `bench`. It would be recommended to rename the target to `secp256k1_bench`.

hebasto commented on src/tests_exhaustive.c:
```diff
@@ -4,6 +4,8 @@
  ***********************************************************************/
 
+#define _POSIX_C_SOURCE 199309L /* for clock_gettime() */
+
```
This doesn’t seem necessary here.

hebasto commented on src/tests.c:
```diff
@@ -4,6 +4,8 @@
  ***********************************************************************/
 
+#define _POSIX_C_SOURCE 199309L /* for clock_gettime() */
+
```
That's annoying that test sources have to be modified. See https://github.com/bitcoin-core/secp256k1/pull/1734#issuecomment-3447767148.

hebasto commented:
https://github.com/bitcoin-core/secp256k1/pull/1732#issuecomment-3335074974:
> This seems overcomplicated and convoluted.

https://github.com/bitcoin-core/secp256k1/pull/1732#discussion_r2394557478:
> This looks compilicated, almost overcomplicated, without much benefits...

The PR author appears to favor simplicity over correctness (see the comment above). I disagree with that approach.

> but you're welcome to provide a commit and I'll see if it should be cherrypicked.

Sure. Here is a branch that addresses some of @purpleKarrot's comments and most of mine: https://github.com/hebasto/secp256k1/commits/251026-pr1732.alt/. However, it doesn't yet include the necessary changes to the Autotools build system.

Raimo33 commented:
I agree, droppped the 3 unrelated commits from this PR, removed the unnecessary `#define`, cherry-picked your solution for the library detection

Raimo33 commented:
rebased and applied @hebasto Makefile fixes
